test/constrained-selection-high-capacity.test.ts

```diff
@@ -0,0 +1,336 @@
+const TEXTURE_SYNC_ID = 'textureSyncTask';
+
+/**
+ * Tests produce visible artifacts for debugging.
+ */
+const article = createArticle();
+document.body.appendChild(article);
+
+describe('constrained selection - high capacity', () => {
+  // Create a <section> for storing visible artifacts.
+  const {section, content} =
+      createSection('constrained selection - high capacity');
+  article.appendChild(section);
+
+  let container: HTMLDivElement;
+  let heading: HTMLHeadingElement;
+  let timingFunctionsShim: TimingFunctionsShim;
+  let scene: Scene;
+  let workScheduler: WorkScheduler;
+  let sampler: Sampler;
+
+  beforeEach(() => {
+    const subsection = document.createElement('div');
+    heading = document.createElement('h3');
+    subsection.appendChild(heading);
+    const subsectionContent = document.createElement('div');
+    subsectionContent.style.display = 'flex';
+    subsectionContent.style.flexDirection = 'row';
+    subsectionContent.style.flexWrap = 'wrap';
+    subsection.appendChild(subsectionContent);
+    content.appendChild(subsection);
+
+    // Create a container <div> of fixed size for the Scene to render into.
+    container = document.createElement('div');
+    container.style.width = '100px';
+    container.style.height = '100px';
+    subsectionContent.appendChild(container);
+
+    timingFunctionsShim = new TimingFunctionsShim();
+    timingFunctionsShim.totalElapsedTimeMs = 1000;
+
+    scene = new Scene({
+      antialiasingFactor: 0,  // Disable antialiasing for pixel-perfect tests.
+      container,
+      defaultTransitionTimeMs: 0,
+      desiredSpriteCapacity: 100000,  // High capacity tests range operations.
+      timingFunctions: timingFunctionsShim,
+    });
+
+    workScheduler = scene[SceneInternalSymbol].workScheduler;
+
+    // Setting maxWorkTimeMs to 0 ensures that only one work task will be
+    // invoked each animation frame. When a task finishes, the WorkScheduler
+    // checks to see if there's any time remaining for more tasks to run, and
+    // always finds that there is none.
+    workScheduler.maxWorkTimeMs = 0;
+
+    // Some long-running tasks do not check remaining time on each iteration of
+    // their internal loops. This cuts down on the frequency and number of calls
+    // to Date.now(). For this test however, we want to constrain the amount of
+    // work that can occur each frame, and so we explicitly set the
+    // stepsBetweenRemainingTimeChecks to 1 so that the amount of remaining time
+    // is checked after each iteration of every loop.
+    scene[SceneInternalSymbol].stepsBetweenRemainingTimeChecks = 1;
+
+    sampler = new Sampler(scene, subsectionContent);
+  });
+
+  afterEach(() => {
+    scene[SceneInternalSymbol].regl.destroy();
+  });
+
+  it('should draw and remove a grid of sprites', async () => {
+    heading.textContent = 'grid of sprites';
+
+    // Draw (initial).
+    expect(workScheduler.taskIds).toEqual([DRAW_TASK_ID]);
```

Comment:
Explicitly testing each task that's run by the WorkScheduler may be overkill. Minor changes to pipeline code will cause this kind of code to fail. But on the other hand, changes to the pipeline code ought to be infrequent. So having explicit, deep integration testing may be worth the price.